src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs

```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
-                        MetricSettings metricSettings = properties.Metrics.FirstOrDefault(x => TimeSpan.Equals(x.TimeGrain, timegrain));
-
-                        if (metricSettings == null)
-                        {
-                            throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Metric timegrain '{0}' is not available", timegrainString));
-                        }
-                        metricSettings.Enabled = this.Enabled;
-                    }
+                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Log category '{0}' is not available", category));
                 }
+
+                logSettings.Enabled = this.Enabled.Value;
             }
+        }
 
-            if (this.RetentionEnabled.HasValue)
+        private void SetAllCategoriesAndTimegrains(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
             {
-                var retentionPolicy = new RetentionPolicy
-                {
-                    Enabled = this.RetentionEnabled.Value,
-                    Days = this.RetentionInDays.Value
-                };
+                return;
+            }
 
-                if (properties.Logs != null)
+            foreach (var log in properties.Logs)
+            {
+                log.Enabled = this.Enabled.Value;
+            }
+
+            foreach (var metric in properties.Metrics)
+            {
+                metric.Enabled = this.Enabled.Value;
+            }
+        }
+
+        private void SetWorkspace(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!string.IsNullOrWhiteSpace(this.WorkspaceId))
+            {
+                if (this.DisableWorkspace)
                 {
-                    foreach (LogSettings logSettings in properties.Logs)
-                    {
-                        logSettings.RetentionPolicy = retentionPolicy;
-                    }
+                    throw new ArgumentException("WorkspaceId and DisableWorkspace cannot be both present.");
                 }
 
-                if (properties.Metrics != null)
+                properties.WorkspaceId = this.WorkspaceId;
+            }
+            else if (this.DisableWorkspace)
+            {
+                properties.WorkspaceId = null;
+            }
+        }
+
+        private void SetServiceBus(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!string.IsNullOrWhiteSpace(this.ServiceBusRuleId))
+            {
+                if (this.DisableServiceBus)
                 {
-                    foreach (MetricSettings metricSettings in properties.Metrics)
-                    {
-                        metricSettings.RetentionPolicy = retentionPolicy;
-                    }
+                    throw new ArgumentException("ServiceBusId and DisableServiceBus cannot be both present.");
```

Comment:
@vcanaa use parameter sets to ensure that these parameters are mutually exclusive